### DISCUSSION TITLE
"winget upgrade" version numbers are wrong

### DISCUSSION THREAD

**Discussion by ourichermath** [microsoft/winget-cli#2644] [Q&A] [answered]
I just reverted from Windows 11 22H2 back to 22H1 using Settings > System > Recovery, and then, I think, Go Back (that's from memory). Now, when I do a "winget upgrade", the resulting list of packages with an update available contains quite a few with incorrect version numbers (both installed and available).  I don't know if this has been cause by reverting or by something I've done.  But, from the list below:

7-zip says the installed version is 22.01, winget says it's 22.00.
Discord doesn't have a version number that I can find, but no update is available when I start it up.
HWiNFO64 says the installed version is 7.30, winget says 7.26
Notepadd++ says the installed version is 8.4.6, winget says 8.4.4
Shotcut has a new version of 22.10.25 and winget says 22.09.23 (I assume it's just not available in winget yet))
Java says the installed version is 8.0_351-b10, winget says 8.0.3310.9 and says 8.0.3410.10 is available
GOG Galaxy says my current version is unknown, which is reasonable since I can't find it in the program. 
PowerShell says 7.2.7 right at the top of the Terminal window where winget says I'm on 7.2.6.0.

Is there any way to clean this up?  Could it be that I've installed the programs from various locations (Microsoft Store, program sites and winget) and that's causing the discrepancy?  Could multiple versions be installed based on where I got them (see previous list)?

```
 winget upgrade --include-unknown                          pwsh  14:47:37 
Name                                               Id                                Version       Available     Source
-----------------------------------------------------------------------------------------------------------------------
7-Zip 22.00 (x64)                                  7zip.7zip                         22.00         22.01         winget
Discord                                            Discord.Discord                   1.0.9005      1.0.9006      winget
HWiNFO64 Version 7.26                              REALiX.HWiNFO                     7.26          7.30          winget
Notepad++ (64-bit x64)                             Notepad++.Notepad++               8.4.4         8.4.6         winget
OBS Studio                                         OBSProject.OBSStudio              27.2.4        28.0.3        winget
Oh My Posh version 8.6.0                           JanDeDobbeleer.OhMyPosh           8.6.0         12.8.0        winget
Shotcut                                            Meltytech.Shotcut                 22.04.25      22.09.23      winget
Java 8 Update 331 (64-bit)                         Oracle.JavaRuntimeEnvironment     8.0.3310.9    8.0.3410.10   winget
Microsoft Visual C++ 2015-2022 Redistributable (x… Microsoft.VCRedist.2015+.x64      14.32.31332.0 14.34.31823.3 winget
Logitech G HUB                                     Logitech.GHUB                     2022.9.315009 2022.10.6382  winget
GOG GALAXY                                         GOG.Galaxy                        Unknown       2.0.55.99     winget
LibreOffice 7.3.4.2                                TheDocumentFoundation.LibreOffice 7.3.4.2       7.4.1.2       winget
GitHub CLI                                         GitHub.cli                        2.13.0        2.18.1        winget
Microsoft Visual C++ 2015-2022 Redistributable (x… Microsoft.VCRedist.2015+.x86      14.32.31332.0 14.34.31823.3 winget
PowerShell 7.2.6.0-x64                             Microsoft.PowerShell              7.2.6.0       7.2.7.0       winget
15 upgrades available.
```

**Comment by ourichermath** [comment, marked as answer]
Let me close this off with what I did.  I spent the day going through that list and trying to upgrade/uninstall all those wrong versions.  They aren't just textual mistakes on winget's part.  For whatever reason (probably a product of the 22H2/21H2 rollback), Windows thought I actually had those versions (the list of Apps in Settings even said so).  So, those apps couldn't be updated because doing so required they be deleted which required an installer for a version that no longer existed.

So, in general, I ended up hunting across the internet for the versions that Windows thought I had.  Github programs weren't bad at all (they maintain archives of old versions).  Some others, I had my original installation program that matched.  The Visual C++ Redistributable's were a pain because I couldn't find those particular versions and ended up manually removing.  Java was horrible.  Java.com looks like it hasn't been updated in years (the last OS it talks about is Windows 10).  Their JavaUninstallTool doesn't even run, let alone do anything.  The fishpoops at Oracle have old versions (if you're smart enough to figure out where they are), but they're behind a registration wall.  And not just a simple name/email one.  They want every kind of information they can think of.  After hours of work, I finally set up a fake registration, downloaded what I needed, installed it and immediately uninstalled it.  If I can, I certainly won't be using Java again.

And, at the Windows level, this kind of thing is a total mess.  If Windows somehow forgets what version of a program you have, you're pretty much rotated.  No way forward, no way backward, no way off.  That shouldn't happen.  The only thing left in my "winget upgrade" list is GOG Galaxy with it's lack of version number (and I've seen bug reports on that going back years and the geniuses over there just continue using a non-standard version number location).  But, since it's still upgradeable and Unknown versions don't appear in the list without that --include-unknown flag, I'm fine with that.

**Comment by denelon** [comment]
What a terrible experience 😔.

I'll try to explain what WinGet is doing and how some of these scenarios happen for different packages.

WinGet looks in the registry for the "displayName" and "displayVersion" values to show what package is installed and what version is installed (sometimes a "marking" version will be displayed instead). The values in the registry are written by the package installers themselves. WinGet doesn't modify these values (with the exception of portable packages where WinGet acts as a proxy "installer").

WinGet looks at those values and some other registry keys like "productCode" and "upgradeCode" to see if there is a matching manifest in one of the configured sources. If there is a match with a manifest, the value in the "appsAndFeatures" is used for comparison to determine if an upgrade is available. In some cases, there may be a "marketing" version ("packageVersion") in the manifest so that value would get displayed to the user instead.

Many MSI and EXE based installers don't report values which is where you encounter the "Unknown" results from. In many cases, the values aren't correctly modified during an upgrade, or the marketing version isn't modified even though the "displayVersion" is. There is also another source of trouble where a "file" version is present, and it's not conforming either to the marketing version or the "displayVersion".

We're trying to meet developers where they are, and to make a best attempt to provide a rational default experience. Sometimes it just doesn't go well. We've still got plenty of heavy lifting to do in WinGet, but at the end of the day, the installer gets to decide what values are reported, and we need to pick the default heuristics to apply. 

Thank you so much for taking the time to provide feedback about your experience.
